internal/nri/plugin.go

```diff
@@ -0,0 +1,385 @@
+// CPU and GPU NUMA nodes. It initializes NVML and returns a fully configured plugin.
+func NewNUMAFilterPlugin(config *spec.Config) (*NUMAFilterPlugin, error) {
+	// Initialize the NVML library
+	nvmlLib := nvml.New()
+	if ret := nvmlLib.Init(); ret != nvml.SUCCESS {
+		return nil, fmt.Errorf("failed to initialize NVML: %v", ret)
+	}
+
+	// Create the nvlib device interface
+	nvlib := device.New(nvmlLib)
+
+	plugin := &NUMAFilterPlugin{
+		nvml:               nvmlLib,
+		nvlib:              nvlib,
+		devicePluginConfig: config,
+	}
+
+	// Get all system NUMA nodes
+	systemNodes, err := plugin.getAllSystemNUMANodes()
+	if err != nil {
+		nvmlLib.Shutdown()
+		return nil, fmt.Errorf("failed to get system NUMA nodes: %w", err)
+	}
+
+	// Get device NUMA nodes
+	deviceNodes, err := plugin.getAllDeviceNUMANodes()
+	if err != nil {
+		nvmlLib.Shutdown()
+		return nil, fmt.Errorf("failed to get device NUMA nodes: %w", err)
+	}
+
+	plugin.cache = &numaTopologyCache{
+		systemNodes: systemNodes,
+		deviceNodes: deviceNodes,
+	}
+
+	return plugin, nil
+}
+
+// Start starts the NRI plugin
+func (p *NUMAFilterPlugin) Start(ctx context.Context) error {
+	// Create the plugin stub with options
+	opts := []nriplugin.Option{
+		nriplugin.WithPluginName(NUMAFilterPluginName),
+		nriplugin.WithPluginIdx("00"), // Use index 00 to run before other plugins
+	}
+
+	s, err := nriplugin.New(p, opts...)
+	if err != nil {
+		return fmt.Errorf("failed to create plugin stub: %v", err)
+	}
+	p.stub = s
+
+	// Start the plugin
+	if err := s.Start(ctx); err != nil {
+		return fmt.Errorf("failed to start plugin: %v", err)
+	}
+
+	return nil
+}
+
+// Stop stops the NRI plugin
+func (p *NUMAFilterPlugin) Stop() {
+	if p.stub != nil {
+		p.stub.Stop()
+	}
+	if p.nvml != nil {
+		p.nvml.Shutdown()
+	}
+}
+
+// CreateContainer is called when a new container is being created. It ensures the container
+// is placed on the correct NUMA nodes based on its assigned GPUs and MIG devices.
+func (p *NUMAFilterPlugin) CreateContainer(ctx context.Context, pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+	nodes, err := p.filterContainerNUMANodes(container)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get NUMA nodes: %w", err)
+	}
+
+	var adjust *api.ContainerAdjustment
+	if len(nodes) > 0 {
+		adjust = &api.ContainerAdjustment{
+			Linux: &api.LinuxContainerAdjustment{
+				Resources: &api.LinuxResources{
+					Cpu: &api.LinuxCPU{
+						Mems: strings.Join(nodes, ","),
+					},
+				},
+			},
+		}
+	}
+
+	return adjust, nil, nil
+}
+
+// UpdateContainer is called when a container's resources are being updated. It ensures the
+// container remains on the correct NUMA nodes based on its assigned GPUs and MIG devices.
+func (p *NUMAFilterPlugin) UpdateContainer(ctx context.Context, pod *api.PodSandbox, container *api.Container, resources *api.LinuxResources) ([]*api.ContainerUpdate, error) {
```

Comment:
From were is this called? Ah... I assume this is the NRI lib that is calling into this.